### PR TITLE
Added GSF Identity Link

### LINKED ISSUE #191: Add identity link for GSF model
This is to enable comparisons with other similar work

### PR DESCRIPTION
**Pull request by gowerc**
Closes #191 

Apologies this PR is a little more involved. Key components are:

- Added Identity Link function for GSF model
- Updated design documentation to include details of the GSF model
- Streamlined parts of the simulation code, I don't think anything was actually fixed / changed I just re-ordered it and made it (slightly) easier to follow and debug
- Added a full re-creation of the Kerioui (2019) model (as they were using the Identity Link) which recovers their results (albeit with worse diagnostics than they report)
- Added testing framework for testing stan functions (needed this to ensure the link identity function was working as intended)
- Misc changes to the manual test files under `design/tests/`

### CODE REVIEW

File: R/LinkGSF.R [hunk 1]
```diff
@@ -182,3 +182,34 @@ link_gsf_dsld <- function(
         contribution_fname = "link_dsld_contribution"
     )
 }
+
+
+# link_gsf_identity-class ----
+
+#' `link_gsf_identity`
+#'
+#' This class extends the general [`link_gsf_abstract`] for the identity of the
```
Comment (danielinteractive) [inline R/LinkGSF.R:191]: Similar comment here on the markdown syntax
```suggestion
#' This class extends the general [link_gsf_abstract()] for the identity of the
```

File: tests/testthat/test-simulations.R [hunk 2]
```diff
@@ -24,7 +26,7 @@ test_that("sim_lm_gsf works as expected", {
     set.seed(123)
     fun_result <- result(df)
     expect_s3_class(fun_result, "data.frame")
-    expect_identical(names(fun_result), c("pt", "time", "sld", "log_haz_link", "study", "arm"))
+    expect_true(all(c("pt", "time", "sld", "log_haz_link", "study", "arm") %in% names(fun_result)))
```
Comment (danielinteractive) [inline tests/testthat/test-simulations.R:29]: I think better to test what exactly you want in the names, not just subset

File: tests/testthat/test-simulations.R [hunk 3]
```diff
@@ -129,3 +136,65 @@ test_that("simulate_joint_data leads to valid DataJoint with almost only default
...
+        beta_cat = c("A" = 0, "B" = -0.6, "C" = 0.6),
+        lm_fun = sim_lm_random_slope(phi = 0),  # No link
+        os_fun = sim_os_weibull(lambda = lambda_real, gamma = gamma_real),
+        .silent = TRUE
+    )
+    osdat <- sim_data$os
+    mod <- survreg(Surv(time, event) ~ cov_cont + cov_cat, data = osdat, dist = "weibull")
+    scale <- mod$scale
+
+    ## Check covariate coeficients
```
Comment (danielinteractive) [inline tests/testthat/test-simulations.R:180]:
```suggestion
    ## Check covariate coefficients
```

**Review by danielinteractive — APPROVED** [review]
Yeah it is quite big but looks roughly good to me without going into details

**Review by danielinteractive — APPROVED** [review]

**Comment by ya0** [comment]
This morning I ran the example in `kerioui-replication.R` and got this output
```R
> mp_k@results$summary(compare_vars)
# A tibble: 11 × 10
   variable               mean     median       sd      mad       q5      q95  rhat ess_bulk ess_tail
   <chr>                 <num>      <num>    <num>    <num>    <num>    <num> <num>    <num>    <num>
 1 lm_gsf_mu_bsld[1] 60.0      60.0       4.20     4.44     53.0     66.6     1.00      27.8    101. 
 2 lm_gsf_mu_ks[1]    0.00594   0.00592   0.000822 0.000770  0.00467  0.00738 1.01     176.     320. 
 3 lm_gsf_mu_kg[1]    0.00173   0.00172   0.000201 0.000195  0.00141  0.00207 1.02      54.0     63.0
 4 lm_gsf_mu_phi[1]   0.185     0.183     0.0376   0.0373    0.127    0.250   0.999    229.     223. 
 5 lm_gsf_omega_bsld  0.714     0.712     0.0536   0.0542    0.634    0.808   1.07      31.4     91.0
 6 lm_gsf_omega_ks    0.123     0.109     0.0643   0.0592    0.0458   0.247   1.01     289.     249. 
 7 lm_gsf_omega_kg    0.825     0.820     0.0860   0.0767    0.696    0.977   0.999     99.8    116. 
 8 lm_gsf_omega_phi   1.46      1.44      0.208    0.215     1.17     1.85    0.999    155.     315. 
 9 lm_gsf_tau        -0.000155 -0.0000557 0.000839 0.000788 -0.00166  0.00108 0.998    370.     439. 
10 lm_gsf_sigma       0.158     0.158     0.00379  0.00356   0.152    0.165   1.00     584.     372. 
11 sm_exp_lambda      0.00205   0.00201   0.000314 0.000300  0.00158  0.00258 1.00     454.     333.
```

I would expect `lm_gsf_tau` for the identity link to be around $0.01$ and `sm_exp_lambda` to be in the order of $1/1450 = 0.0006896552..$ (its prior value).

Do you get something similar?

**Comment by ya0** [comment]
@gowerc I saw that you just recently closed this PR. Sorry for not replying sooner. But only today I was experimenting with this identity link branch.

**Comment by ya0** [comment]
The gray line you show here is the prior.

The posterior $\lambda$ mean $0.00205$ looks like a normal value for a survival only fit with no link. or for the ttg/derivative cases. But in case of the identity link it should be smaller (i.e. higher).

On [julia for joint models](https://docs.roche.com/#/julia-joint-models/latest/pages/joint_model.html#joint-models-with-a-link-function-operator) i have documented some outputs for the different cases made with my models. Here is a screenshot
image

The only reference i have is the output of the original Stan/R code that was alongside the paper.
```R
               mean  se_mean        sd n_eff    Rhat
mu_1       62.26464  0.61117   4.48100    54 1.03979
mu_2        0.00487  0.00024   0.00197    65 1.06537
mu_3        0.00181  0.00007   0.00029    20 1.15857
mu_4        0.20024  0.01110   0.08510    59 1.07103
Omega[1]    0.73227  0.00650   0.05196    64 1.07858
Omega[2]    0.56993  0.05826   0.43380    55 1.07908
Omega[3]    0.83742  0.01418   0.11937    71 1.04076
Omega[4]    1.54100  0.02664   0.35360   176 1.00189
sigma       0.17672  0.00026   0.00602   545 1.00097
lambda   1505.44167 13.41165 339.75674   642 0.99912
beta        0.01032  0.00005   0.00134   652 1.00179
```
Here the `lambda` and `beta` parameter were estimated relatively close to the valued used for simulating the dataset.

**Pull request merged**